Add explicit mode for SGX (#432)
iExecBlockchainComputing/iexec-worker: README.md

cd iexec-worker
./gradlew bootRun --refresh-dependencies
```

> #### Configuration

Reviewer: Thanks for upgrading doc :)

You can configure the worker with the following properties:

| Environment variables                         | Description                                                                                                                      | Values           | Default value                                             |
|-----------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------|------------------|-----------------------------------------------------------|

> | IEXEC_WORKER_PORT                             | Port the worker's server will use.                                                                                               | Positive integer | 13100                                                     |

Reviewer: ```suggestion
| IEXEC_WORKER_PORT                             | Server port of the worker .                                                                                              | Positive integer | 13100                                                     |
```

| IEXEC_CORE_PROTOCOL                           | Protocol to connect to the Scheduler.                                                                                            | String           | http                                                      |
| IEXEC_CORE_HOST                               | Host to connect to the Scheduler.                                                                                                | String           | localhost                                                 |
| IEXEC_CORE_PORT                               | Port used by the Scheduler.                                                                                                      | Positive integer | 13000                                                     |

> | IEXEC_WORKER_NAME                             | Name the worker should be known by.                                                                                              | String           | worker                                                    |

Reviewer: ```suggestion
| IEXEC_WORKER_NAME                             | Public name of the worker.                                                                                          | String           | worker                                                    |
```

> | IEXEC_WORKER_BASE_DIR                         | Path to the folder within which the worker will read and write tasks' inputs and outputs.                                        | String           | /tmp/iexec-worker                                         |

Reviewer: ```suggestion
| IEXEC_WORKER_BASE_DIR                         | Path to the folder within which the worker will read-and-write inputs and outputs of tasks.                                        | String           | /tmp/iexec-worker                                         |
```

> | IEXEC_WORKER_OVERRIDE_AVAILABLE_CPU_COUNT     | Number of CPUs to use. If not set, n-1 CPUs will be used, where n is the number of available processors.                         | Positive integer |                                                           |

Reviewer: ```suggestion
| IEXEC_WORKER_OVERRIDE_AVAILABLE_CPU_COUNT     | Number of CPUs available for computing distinct tasks. If not set, n-1 CPUs will be used, where n is the number of available processors.                         | Positive integer |                                                           |
```

> | IEXEC_WORKER_GPU_ENABLED                      | Whether to use the GPU instead of the CPU. Note that if it is true, `IEXEC_WORKER_OVERRIDE_AVAILABLE_CPU_COUNT` will be ignored. | Boolean          | false                                                     |

Reviewer: ```suggestion
| IEXEC_WORKER_GPU_ENABLED                      | Declares if the worker is able to compute tasks requesting GPU mode. Note that if it is true, `IEXEC_WORKER_OVERRIDE_AVAILABLE_CPU_COUNT` will be ignored. | Boolean          | false                                                     |
```

| IEXEC_GAS_PRICE_MULTIPLIER                    | Transactions will be sent with `networkGasPrice * IEXEC_GAS_PRICE_MULTIPLIER`.                                                   | Float            | 1.3                                                       |
| IEXEC_GAS_PRICE_CAP                           | In Wei, will be used for transactions if `networkGasPrice * IEXEC_GAS_PRICE_MULTIPLIER > gasPriceCap`.                           | Integer          | 22000000000                                               |
| IEXEC_WORKER_OVERRIDE_BLOCKCHAIN_NODE_ADDRESS | If set, will be used instead of the address given by the blockchain adapter.                                                     | String           |                                                           |

> | IEXEC_DEVELOPER_LOGGER_ENABLED                | Whether to print apps' logs.                                                                                                     | Boolean          | false                                                     |

Reviewer: ```suggestion
| IEXEC_DEVELOPER_LOGGER_ENABLED                | Whether to print application logs of tasks.                                                                                                     | Boolean          | false                                                     |
```

> | IEXEC_WORKER_TEE_COMPUTE_MAX_HEAP_SIZE_GB     | Max heap size for TEE apps.                                                                                                      | Positive integer | 8                                                         |

Reviewer 2: Do we have minimum requirement for the worker ? (CPU / RAM)
What happens if the provided size is bigger than system available memory ?

Author: I don't actually know. That'll be points we will have to work on.

Reviewer: Current implementation is not optimal but protects workers.
In the future, a worker should expose to the core how much RAM it has hence the Scheduler would be able to schedule tasks on compatible workers.

> | IEXEC_WORKER_DOCKER_NETWORK_NAME              | Name of the network the LAS is in. Useless if worker runs only standard tasks.                                                   | String           | iexec-worker-net                                          |

Reviewer: ```suggestion
| IEXEC_WORKER_DOCKER_NETWORK_NAME              | Internal Docker network name of the worker. Required for communication between worker and launched-by-worker containers.                                                  | String           | iexec-worker-net                                          |
```

> | IEXEC_WORKER_DOCKER_REGISTRY_USERNAME_0       | Username to pull apps from [docker official website](https://hub.docker.com/)                                                    | String           |                                                           |

Reviewer: ```suggestion
| IEXEC_WORKER_DOCKER_REGISTRY_USERNAME_0       | Username to pull apps from [docker official registry](https://hub.docker.com/)                                                    | String           |                                                           |
```

Reviewer: Link is to update too

Author: What would be the official registry?

> | IEXEC_WORKER_DOCKER_REGISTRY_PASSWORD_0       | Password to  pull apps from [docker official website](https://hub.docker.com/)                                                   | String           |                                                           |

Reviewer: ```suggestion
| IEXEC_WORKER_DOCKER_REGISTRY_PASSWORD_0       | Password to  pull apps from [docker official registry](https://hub.docker.com/)                                                   | String           |                                                           |
```

| IEXEC_WORKER_DOCKER_REGISTRY_PASSWORD_1       | Password to pull apps from secondary registry.                                                                                   | String           |                                                           |
| IEXEC_WORKER_WALLET_PATH                      | Path to the wallet that should be used by the worker.                                                                            | String           | ./src/main/resources/wallet/encrypted-wallet_worker1.json |
| IEXEC_WORKER_WALLET_PASSWORD                  | Password of the wallet that should be used by the worker.                                                                        | String           | whatever                                                  |
| IEXEC_SCONE_SHOW_VERSION                      | Whether to display version of Scone.                                                                                             | Boolean          | true                                                      |

> | IEXEC_SCONE_LOG_LEVEL                         | Log level Scone should use.                                                                                                      | # TO COMPLETE #  | debug                                                     |

Author: I haven't found any concrete info on Scone log level so can't complete this part myself.

Reviewer: https://sconedocs.github.io/SCONE_ENV/#debug :
_SCONE_LOG set to value between 0 (no) and 7 (debug) to see more or less messages on stderr from the SCONE platform. Messages could be warnings if certain functions are not (yet) implemented inside of an enclave._
String values alternative got to be somewhere.

Author: Thanks, I'll add that info.

Reviewer 2: Nice work for the properties.
Maybe add that properties are `iexec.core.port` (lowercase with `.`) and can be overriden with environment variables `IEXEC_CORE_PORT` (uppercase with `_`)

Author: That's not always true so it can't be a general rule. But I can add a new column if you find it useful - or simply rename `Property` into `Env var`.

Reviewer 2: Hmm, I don't know. See how the spring boot documentation is written.
https://docs.spring.io/spring-boot/docs/current/reference/html/application-properties.html

Author: I think Spring Boot is different from our project. Spring Boot is a framework which we interact with in our code (= properties). iExec Worker is more like an app people interact with through env vars.

Reviewer 2: I don't understand what you mean by **Spring Boot is different from our project**.
Spring Boot is a framework and we build applications upon it.
The documentation I provided show the available properties defined by the framework and available at the application level, given the related spring boot modules have been enabled/imported.
For further reading on Spring Boot applications properties, you can read:
* https://docs.spring.io/spring-boot/docs/current/reference/html/application-properties.html
* https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config.typesafe-configuration-properties.relaxed-binding
* https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config.typesafe-configuration-properties.relaxed-binding.environment-variables
* https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config

In my understanding, as the environment variables naming convention is rather strict and well adopted, there is a kind of consensus to use another notation for properties.

> | IEXEC_SCONE_LOG_LEVEL                         | Log level Scone should use.                                                                                                      | [0,7] or String  | debug                                                     |

Reviewer 2: With a production mindset, wouldn't it be better to have a default value like `WARN` ?
The fact to have those logs at the `DEBUG` level should be the exception.

Author: That's an eternal debate: whether we want to have production-ready code or we want to force users to customize their workers.

Reviewer: Yes, we need to work on that someday : `Dev configuration vs Production configuration`, we can add that to the backlog.

> | IEXEC_WORKER_SCONTAIN_REGISTRY_NAME           | Name of the registry the LAS should be retrieved from.                                                                           | String           | registry.scontain.com:5050                                |

Reviewer: ```suggestion
| IEXEC_WORKER_SCONTAIN_REGISTRY_NAME           | Name of the Scontain registry. Currently used to pull LAS image.                                                                      | String           | registry.scontain.com:5050                                |
```

> | IEXEC_WORKER_SCONTAIN_REGISTRY_USERNAME       | Username to connect to the registry the LAS should be retrieved from.                                                            | String           |                                                           |

Reviewer: ```suggestion
| IEXEC_WORKER_SCONTAIN_REGISTRY_USERNAME       | Username to connect to the Scontain registry.                                                            | String           |                                                           |
```

> | IEXEC_WORKER_SCONTAIN_REGISTRY_PASSWORD       | Password to connect to the registry the LAS should be retrieved from.                                                            | String           |                                                           |

Reviewer: ```suggestion
| IEXEC_WORKER_SCONTAIN_REGISTRY_PASSWORD       | Password to connect to the Scontain registry.                                                            | String           |                                                           |
```

| IEXEC_LAS_PORT                                | Port the LAS should be started on.                                                                                               | Positive integer | 18766                                                     |
| IEXEC_CORE_GRAYLOG_HOST                       | Graylog host.                                                                                                                    | String           | localhost                                                 |
| IEXEC_CORE_GRAYLOG_PORT                       | Graylog port.                                                                                                                    | Positive integer | 12201                                                     |

> | IEXEC_WORKER_SGX_DRIVER_MODE                  | SGX driver that should be used                                                                                                   | { NONE, LEGACY } | NONE                                                      |

Reviewer: ```suggestion
| IEXEC_WORKER_SGX_DRIVER_MODE                  | Intel® SGX driver that should be used.                                                                                                   | { NONE, LEGACY } | NONE                                                      |
```
